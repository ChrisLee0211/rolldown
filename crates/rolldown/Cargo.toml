[package]
edition = "2021"
name    = "rolldown"
version = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
derivative                   = { workspace = true }
futures                      = { workspace = true }
rolldown_core                = { path = "../rolldown_core" }
rolldown_error               = { path = "../rolldown_error" }
rolldown_plugin              = { path = "../rolldown_plugin" }
rolldown_plugin_node_resolve = { path = "../rolldown_plugin_node_resolve" }
rolldown_tracing             = { path = "../rolldown_tracing" }
sugar_path                   = { workspace = true }
tokio                        = { workspace = true, features = ["full"] }

[dev_dependencies]
criterion                    = "0.4.0"
insta                        = { workspace = true }
rolldown_plugin_node_resolve = { path = "../rolldown_plugin_node_resolve" }
rolldown_test_utils          = { path = "../rolldown_test_utils" }
testing_macros               = { workspace = true }

[target.'cfg(not(target_os = "linux"))'.dev_dependencies]
mimalloc-rust = "0.2"

[target.'cfg(all(target_os = "linux", not(all(target_env = "musl", target_arch = "aarch64"))))'.dev_dependencies]
mimalloc-rust = { version = "0.2", features = ["local-dynamic-tls"] }

[[bench]]
harness = false
name    = "threejs"
