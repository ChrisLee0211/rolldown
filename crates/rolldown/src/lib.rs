mod bundler;
mod input_options;
mod output_options;
pub use {
  bundler::Bundler,
  input_options::{
    default_warning_handler, BuiltinsOptions, InputItem, InputOptions, IsExternal,
    NodeResolveOptions, TsConfig,
  },
  output_options::{ExportMode, FileNameTemplate, ModuleFormat, OutputOptions},
  rolldown_core::{Asset, BuildResult},
};
