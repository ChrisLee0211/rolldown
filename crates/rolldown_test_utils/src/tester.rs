use std::{
  collections::HashSet,
  path::{Path, PathBuf},
  sync::Arc,
};

use futures::FutureExt;
use rolldown_core::BuildError;

use crate::test_config::TestConfig;

pub struct Tester {
  pub config: TestConfig,
  pub warnings: Vec<BuildError>,
}

impl Tester {
  pub fn from_config_path(filepath: &Path) -> Self {
    let test_config = TestConfig::from_config_path(filepath);
    Self {
      config: test_config,
      warnings: Default::default(),
    }
  }

  pub fn input_options(&self, cwd: PathBuf) -> rolldown_core::InputOptions {
    rolldown_core::InputOptions {
      input: self.config.input.input.clone(),
      cwd,
      treeshake: self.config.input.treeshake,
      is_external: {
        let external = self
          .config
          .input
          .external
          .clone()
          .into_iter()
          .collect::<HashSet<_>>();
        Arc::new(move |specifier, _importer, _| {
          let external = external.clone();
          futures::future::ready(Ok(external.contains(specifier))).boxed()
        })
      },
      ..Default::default()
    }
  }
}
