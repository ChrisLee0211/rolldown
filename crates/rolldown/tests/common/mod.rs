use std::{
  path::{Path, PathBuf},
  str::FromStr,
};

use rolldown::Bundler;
use rolldown::{Asset, BuildResult, ExportMode, ModuleFormat, OutputOptions};
use rolldown_test_utils::tester::Tester;

pub struct CompiledFixture {
  pub tester: Tester,
  pub bundler: Bundler,
  pub output: BuildResult<Vec<Asset>>,
  pub dir_name: String,
  pub fixture_path: PathBuf,
}

impl CompiledFixture {
  pub fn output_friendly_to_snapshot(self) -> String {
    if self.tester.config.expect_error {
      let mut errors = self.output.unwrap_err().into_vec();
      errors.sort();
      let snapshot = [format!("---------- ERRORS ----------")]
        .into_iter()
        .chain(errors.iter().map(|err| {
          format!(
            "{}: {}",
            err.kind.code(),
            err.kind.to_readable_string(&self.fixture_path)
          )
        }))
        .collect::<Vec<_>>()
        .join("\n");

      return snapshot;
    }

    let mut assets = self.output.as_ref().unwrap().iter().collect::<Vec<_>>();
    assets.sort_by_key(|c| &c.filename);
    assets
      .iter()
      .flat_map(|asset| {
        [
          format!("---------- {} ----------", asset.filename),
          asset.content.trim().to_string(),
        ]
      })
      .chain(if self.tester.warnings.lock().unwrap().is_empty() {
        vec![]
      } else {
        let mut warnings = self.tester.warnings.lock().unwrap();
        warnings.sort();
        vec![
          format!("---------- WARNINGS ----------"),
          warnings
            .iter()
            .map(|w| {
              format!(
                "{}: {}",
                w.kind.code(),
                w.kind.to_readable_string(&self.fixture_path)
              )
            })
            .collect::<Vec<_>>()
            .join("\n"),
        ]
      })
      .collect::<Vec<_>>()
      .join("\n")
  }
}

pub async fn compile_fixture(test_config_path: &Path) -> CompiledFixture {
  let fixture_path = test_config_path.parent().unwrap();

  let tester = Tester::from_config_path(test_config_path);

  let mut bundler = Bundler::new(tester.input_options(fixture_path.to_path_buf()));

  if fixture_path.join("dist").is_dir() {
    std::fs::remove_dir_all(fixture_path.join("dist")).unwrap();
  }

  let output = bundler
    .generate(OutputOptions {
      // dir: Some(fixture_path.join("dist").to_string_lossy().to_string()),
      format: ModuleFormat::from_str(&tester.config.output.format).unwrap(),
      export_mode: ExportMode::from_str(&tester.config.output.export_mode).unwrap(),
      ..Default::default()
    })
    .await;
  let fixture_name = fixture_path
    .file_name()
    .unwrap()
    .to_string_lossy()
    .to_string();

  CompiledFixture {
    tester,
    bundler,
    output,
    dir_name: fixture_name,
    fixture_path: fixture_path.to_path_buf(),
  }
}

pub fn run_test(test_config_path: &Path) {
  // compile the fixture folder
  let compiled_fx = tokio::runtime::Runtime::new()
    .unwrap()
    .block_on(crate::common::compile_fixture(test_config_path));

  // If the test config has an expected error, assert that the error matches
  if let Some(expected_error) = compiled_fx.tester.config.expected_error {
    let errors = compiled_fx
      .output
      .expect_err("Expected error but got success")
      .into_vec();

    // TODO: should supports multiple errors
    let error = errors
      .iter()
      .find(|error| {
        error.kind.code() == expected_error.code
          && error.kind.to_readable_string(&compiled_fx.fixture_path) == expected_error.message
      })
      .unwrap_or_else(|| {
        eprintln!("errors: {errors:#?}");
        panic!("should find the error");
      });
    assert_eq!(error.kind.code(), expected_error.code);
    assert_eq!(
      error.kind.to_readable_string(&compiled_fx.fixture_path),
      expected_error.message
    );
    return;
  }

  // Otherwise, assert that the output matches the snapshot

  // Configure insta to use the test config path as the snapshot path
  let fixture_folder = test_config_path.parent().unwrap();
  let mut settings = insta::Settings::clone_current();
  settings.set_snapshot_path(fixture_folder);
  settings.set_prepend_module_to_snapshot(false);
  settings.set_input_file(fixture_folder);
  settings.bind(|| {
    insta::assert_snapshot!("output", compiled_fx.output_friendly_to_snapshot());
  });
}
