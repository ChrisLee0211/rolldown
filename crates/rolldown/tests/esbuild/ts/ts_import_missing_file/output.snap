---
source: crates/rolldown/tests/common/mod.rs
expression: compiled_fx.output_friendly_to_snapshot()
input_file: crates/rolldown/tests/esbuild/ts/type_script_decorators_keep_names
---
---------- main.js ----------
// main.ts
function _arrayLikeToArray(arr, len) {
    (null == len || len > arr.length) && (len = arr.length);
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _arrayWithHoles(arr) {
    if (Array.isArray(arr)) return arr;
}
function _decorate(decorators, factory, superClass) {
    var r = factory(function initialize(O) {
        _initializeInstanceElements(O, decorated.elements);
    }, superClass);
    var decorated = _decorateClass(_coalesceClassElements(r.d.map(_createElementDescriptor)), decorators);
    _initializeClassElements(r.F, decorated.elements);
    return _runClassFinishers(r.F, decorated.finishers);
}
function _createElementDescriptor(def) {
    var key = _toPropertyKey(def.key);
    var descriptor;
    if ("method" === def.kind) {
        descriptor = {
            value: def.value,
            writable: true,
            configurable: true,
            enumerable: false
        };
        Object.defineProperty(def.value, "name", {
            value: "symbol" === _typeof(key) ? "" : key,
            configurable: true
        });
    } else "get" === def.kind ? descriptor = {
        get: def.value,
        configurable: true,
        enumerable: false
    } : "set" === def.kind ? descriptor = {
        set: def.value,
        configurable: true,
        enumerable: false
    } : "field" === def.kind && (descriptor = {
        configurable: true,
        writable: true,
        enumerable: true
    });
    var element = {
        kind: "field" === def.kind ? "field" : "method",
        key: key,
        placement: def.static ? "static" : "field" === def.kind ? "own" : "prototype",
        descriptor: descriptor
    };
    def.decorators && (element.decorators = def.decorators);
    "field" === def.kind && (element.initializer = def.value);
    return element;
}
function _coalesceGetterSetter(element, other) {
    void 0 !== element.descriptor.get ? other.descriptor.get = element.descriptor.get : other.descriptor.set = element.descriptor.set;
}
function _coalesceClassElements(elements) {
    var newElements = [];
    var isSameElement = function(other) {
        return "method" === other.kind && other.key === element.key && other.placement === element.placement;
    };
    for(var i = 0; i < elements.length; i++){
        var element = elements[i];
        var other;
        if ("method" === element.kind && (other = newElements.find(isSameElement))) {
            if (_isDataDescriptor(element.descriptor) || _isDataDescriptor(other.descriptor)) {
                if (_hasDecorators(element) || _hasDecorators(other)) throw new ReferenceError("Duplicated methods (" + element.key + ") can't be decorated.");
                other.descriptor = element.descriptor;
            } else {
                if (_hasDecorators(element)) {
                    if (_hasDecorators(other)) throw new ReferenceError("Decorators can't be placed on different accessors with for the same property (" + element.key + ").");
                    other.decorators = element.decorators;
                }
                _coalesceGetterSetter(element, other);
            }
        } else newElements.push(element);
    }
    return newElements;
}
function _hasDecorators(element) {
    return element.decorators && element.decorators.length;
}
function _isDataDescriptor(desc) {
    return void 0 !== desc && !(void 0 === desc.value && void 0 === desc.writable);
}
function _initializeClassElements(F, elements) {
    var proto = F.prototype;
    [
        "method",
        "field"
    ].forEach(function(kind) {
        elements.forEach(function(element) {
            var placement = element.placement;
            if (element.kind === kind && ("static" === placement || "prototype" === placement)) {
                var receiver = "static" === placement ? F : proto;
                _defineClassElement(receiver, element);
            }
        });
    });
}
function _initializeInstanceElements(O, elements) {
    [
        "method",
        "field"
    ].forEach(function(kind) {
        elements.forEach(function(element) {
            element.kind === kind && "own" === element.placement && _defineClassElement(O, element);
        });
    });
}
function _defineClassElement(receiver, element) {
    var descriptor = element.descriptor;
    if ("field" === element.kind) {
        var initializer = element.initializer;
        descriptor = {
            enumerable: descriptor.enumerable,
            writable: descriptor.writable,
            configurable: descriptor.configurable,
            value: void 0 === initializer ? void 0 : initializer.call(receiver)
        };
    }
    Object.defineProperty(receiver, element.key, descriptor);
}
function _decorateClass(elements, decorators) {
    var newElements = [];
    var finishers = [];
    var placements = {
        static: [],
        prototype: [],
        own: []
    };
    elements.forEach(function(element) {
        _addElementPlacement(element, placements);
    });
    elements.forEach(function(element) {
        if (!_hasDecorators(element)) return newElements.push(element);
        var elementFinishersExtras = _decorateElement(element, placements);
        newElements.push(elementFinishersExtras.element);
        newElements.push.apply(newElements, elementFinishersExtras.extras);
        finishers.push.apply(finishers, elementFinishersExtras.finishers);
    });
    if (!decorators) return {
        elements: newElements,
        finishers: finishers
    };
    var result = _decorateConstructor(newElements, decorators);
    finishers.push.apply(finishers, result.finishers);
    result.finishers = finishers;
    return result;
}
function _addElementPlacement(element, placements, silent) {
    var keys = placements[element.placement];
    if (!silent && -1 !== keys.indexOf(element.key)) throw new TypeError("Duplicated element (" + element.key + ")");
    keys.push(element.key);
}
function _decorateElement(element, placements) {
    var extras = [];
    var finishers = [];
    for(var decorators = element.decorators, i = decorators.length - 1; i >= 0; i--){
        var keys = placements[element.placement];
        keys.splice(keys.indexOf(element.key), 1);
        var elementObject = _fromElementDescriptor(element);
        var elementFinisherExtras = _toElementFinisherExtras((0, decorators[i])(elementObject) || elementObject);
        element = elementFinisherExtras.element;
        _addElementPlacement(element, placements);
        elementFinisherExtras.finisher && finishers.push(elementFinisherExtras.finisher);
        var newExtras = elementFinisherExtras.extras;
        if (newExtras) {
            for(var j = 0; j < newExtras.length; j++)_addElementPlacement(newExtras[j], placements);
            extras.push.apply(extras, newExtras);
        }
    }
    return {
        element: element,
        finishers: finishers,
        extras: extras
    };
}
function _decorateConstructor(elements, decorators) {
    var finishers = [];
    for(var i = decorators.length - 1; i >= 0; i--){
        var obj = _fromClassDescriptor(elements);
        var elementsAndFinisher = _toClassDescriptor((0, decorators[i])(obj) || obj);
        void 0 !== elementsAndFinisher.finisher && finishers.push(elementsAndFinisher.finisher);
        if (void 0 !== elementsAndFinisher.elements) {
            elements = elementsAndFinisher.elements;
            for(var j = 0; j < elements.length - 1; j++)for(var k = j + 1; k < elements.length; k++)if (elements[j].key === elements[k].key && elements[j].placement === elements[k].placement) throw new TypeError("Duplicated element (" + elements[j].key + ")");
        }
    }
    return {
        elements: elements,
        finishers: finishers
    };
}
function _fromElementDescriptor(element) {
    var obj = {
        kind: element.kind,
        key: element.key,
        placement: element.placement,
        descriptor: element.descriptor
    };
    var desc = {
        value: "Descriptor",
        configurable: true
    };
    Object.defineProperty(obj, Symbol.toStringTag, desc);
    "field" === element.kind && (obj.initializer = element.initializer);
    return obj;
}
function _toElementDescriptors(elementObjects) {
    if (void 0 === elementObjects) return;
    return _toArray(elementObjects).map(function(elementObject) {
        var element = _toElementDescriptor(elementObject);
        _disallowProperty(elementObject, "finisher", "An element descriptor");
        _disallowProperty(elementObject, "extras", "An element descriptor");
        return element;
    });
}
function _toElementDescriptor(elementObject) {
    var kind = String(elementObject.kind);
    if ("method" !== kind && "field" !== kind) throw new TypeError('An element descriptor\'s .kind property must be either "method" or "field", but a decorator created an element descriptor with .kind "' + kind + '"');
    var key = _toPropertyKey(elementObject.key);
    var placement = String(elementObject.placement);
    if ("static" !== placement && "prototype" !== placement && "own" !== placement) throw new TypeError('An element descriptor\'s .placement property must be one of "static", "prototype" or "own", but a decorator created an element descriptor with .placement "' + placement + '"');
    var descriptor = elementObject.descriptor;
    _disallowProperty(elementObject, "elements", "An element descriptor");
    var element = {
        kind: kind,
        key: key,
        placement: placement,
        descriptor: Object.assign({}, descriptor)
    };
    if ("field" !== kind) _disallowProperty(elementObject, "initializer", "A method descriptor");
    else {
        _disallowProperty(descriptor, "get", "The property descriptor of a field descriptor");
        _disallowProperty(descriptor, "set", "The property descriptor of a field descriptor");
        _disallowProperty(descriptor, "value", "The property descriptor of a field descriptor");
        element.initializer = elementObject.initializer;
    }
    return element;
}
function _toElementFinisherExtras(elementObject) {
    var element = _toElementDescriptor(elementObject);
    var finisher = _optionalCallableProperty(elementObject, "finisher");
    var extras = _toElementDescriptors(elementObject.extras);
    return {
        element: element,
        finisher: finisher,
        extras: extras
    };
}
function _fromClassDescriptor(elements) {
    var obj = {
        kind: "class",
        elements: elements.map(_fromElementDescriptor)
    };
    var desc = {
        value: "Descriptor",
        configurable: true
    };
    Object.defineProperty(obj, Symbol.toStringTag, desc);
    return obj;
}
function _toClassDescriptor(obj) {
    var kind = String(obj.kind);
    if ("class" !== kind) throw new TypeError('A class descriptor\'s .kind property must be "class", but a decorator created a class descriptor with .kind "' + kind + '"');
    _disallowProperty(obj, "key", "A class descriptor");
    _disallowProperty(obj, "placement", "A class descriptor");
    _disallowProperty(obj, "descriptor", "A class descriptor");
    _disallowProperty(obj, "initializer", "A class descriptor");
    _disallowProperty(obj, "extras", "A class descriptor");
    var finisher = _optionalCallableProperty(obj, "finisher");
    var elements = _toElementDescriptors(obj.elements);
    return {
        elements: elements,
        finisher: finisher
    };
}
function _disallowProperty(obj, name, objectType) {
    if (void 0 !== obj[name]) throw new TypeError(objectType + " can't have a ." + name + " property.");
}
function _optionalCallableProperty(obj, name) {
    var value = obj[name];
    if (void 0 !== value && "function" != typeof value) throw new TypeError("Expected '" + name + "' to be a function");
    return value;
}
function _runClassFinishers(constructor, finishers) {
    for(var i = 0; i < finishers.length; i++){
        var newConstructor = (0, finishers[i])(constructor);
        if (void 0 !== newConstructor) {
            if ("function" != typeof newConstructor) throw new TypeError("Finishers must return a constructor.");
            constructor = newConstructor;
        }
    }
    return constructor;
}
function _iterableToArray(iter) {
    if ("undefined" != typeof Symbol && null != iter[Symbol.iterator] || null != iter["@@iterator"]) return Array.from(iter);
}
function _nonIterableRest() {
    throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _toArray(arr) {
    return _arrayWithHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableRest();
}
function _toPrimitive(input, hint) {
    if ("object" !== _typeof(input) || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" !== _typeof(res)) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function _toPropertyKey(arg) {
    var key = _toPrimitive(arg, "string");
    return "symbol" === _typeof(key) ? key : String(key);
}
var _typeof = function(obj) {
    return obj && "undefined" != typeof Symbol && obj.constructor === Symbol ? "symbol" : typeof obj;
};
function _unsupportedIterableToArray(o, minLen) {
    if (!o) return;
    if ("string" == typeof o) return _arrayLikeToArray(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    "Object" === n && o.constructor && (n = o.constructor.name);
    if ("Map" === n || "Set" === n) return Array.from(n);
    if ("Arguments" === n || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}
_decorate([
    decoratorMustComeAfterName
], function(_initialize) {
    class Foo$1 {
        static name = "Foo";
        constructor(){
            _initialize(this);
        }
    }
    return {
        F: Foo$1,
        d: []
    };
});
